cpp/src/io/csv/reader_impl.cu

```diff
@@ -479,43 +506,43 @@ void reader::impl::gather_row_offsets(host_span<char const> const data,
         if (rows_out_of_range != 0) {
           // Keep one row out of range (used to infer length of previous row)
           auto new_row_offsets_size =
-            row_offsets_.size() - std::min(rows_out_of_range - 1, row_offsets_.size());
-          row_offsets_.resize(new_row_offsets_size);
+            all_row_offsets.size() - std::min(rows_out_of_range - 1, all_row_offsets.size());
+          all_row_offsets.resize(new_row_offsets_size, stream);
           // Implies we reached the end of the range
           break;
         }
       }
       // num_rows does not include blank rows
       if (num_rows >= 0) {
-        if (row_offsets_.size() > header_rows + static_cast<size_t>(num_rows)) {
+        if (all_row_offsets.size() > header_rows + static_cast<size_t>(num_rows)) {
           size_t num_blanks =
-            cudf::io::csv::gpu::count_blank_rows(opts.view(), data_, row_offsets_, stream);
-          if (row_offsets_.size() - num_blanks > header_rows + static_cast<size_t>(num_rows)) {
+            cudf::io::csv::gpu::count_blank_rows(opts.view(), d_data, all_row_offsets, stream);
+          if (all_row_offsets.size() - num_blanks > header_rows + static_cast<size_t>(num_rows)) {
             // Got the desired number of rows
             break;
           }
         }
       }
     } else {
       // Discard data (all rows below skip_rows), keeping one character for history
-      size_t discard_bytes = std::max(data_.size(), sizeof(char)) - sizeof(char);
+      size_t discard_bytes = std::max(d_data.size(), sizeof(char)) - sizeof(char);
       if (discard_bytes != 0) {
-        data_.erase(data_.begin(), data_.begin() + discard_bytes);
+        erase_except_last(d_data, stream);
```

Comment:
@jrhemstad says that we should be able to add a factory that constructs `device_uvector` from iterators; somewhat like this:

```suggestion
        d_data = make_uvector(d_data.begin() + discard_bytes, d_data.end(), stream);
```

Reply:
That looks so much cleaner but it's probably not as fast in this case. I really want to keep the capacity here. With skip_rows, the loop is: append 64MB of data -> process -> discard all but last char -> repeat until skip_rows is reached. After this, data is only appended.

Reply:
But then it is resized here: https://github.com/rapidsai/cudf/pull/7805/files#diff-369b6eb245d8516f57edbc1acedfd97e9b52e6b9ef9a600f81447b2de78c0cf3R194

Reply:
Resize keeps the capacity: https://docs.rapids.ai/api/librmm/stable/classrmm_1_1device__uvector.html#a12352ccb04d368540eb7ebf0ceb6ef20
Otherwise I would have to implement a special container for the device data here.